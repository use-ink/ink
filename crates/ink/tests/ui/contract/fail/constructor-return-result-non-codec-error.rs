#![allow(unexpected_cfgs)]

#[ink::contract]
mod contract {
    #[ink(storage)]
    pub struct Contract {}

    #[derive(Debug, PartialEq, Eq)]
    pub enum Error {
        Foo,
    }

    impl Contract {
        #[ink(constructor)]
        pub fn constructor() -> Result<Self, Error> {
            Ok(Self {})
        }

        #[ink(message)]
        pub fn message(&self) {}
    }
}

fn main() {}
