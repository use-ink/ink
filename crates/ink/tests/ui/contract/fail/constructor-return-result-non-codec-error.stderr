error[E0277]: the trait bound `Result<Result<(), &contract::Error>, LangError>: Encode` is not satisfied
  --> tests/ui/contract/fail/constructor-return-result-non-codec-error.rs:13:9
   |
13 | /         pub fn constructor() -> Result<Self, Error> {
14 | |             Ok(Self {})
15 | |         }
   | |_________^ the trait `Encode` is not implemented for `Result<Result<(), &contract::Error>, LangError>`
   |
   = help: the trait `Encode` is implemented for `Result<T, E>`
note: required by a bound in `return_value`
  --> $WORKSPACE/crates/env/src/api.rs
   |
   | pub fn return_value<R>(return_flags: ReturnFlags, return_value: &R) -> !
   |        ------------ required by a bound in this function
   | where
   |     R: scale::Encode,
   |        ^^^^^^^^^^^^^ required by this bound in `return_value`

error[E0277]: the trait bound `contract::Error: WrapperTypeDecode` is not satisfied
  --> tests/ui/contract/fail/constructor-return-result-non-codec-error.rs:13:33
   |
13 |           pub fn constructor() -> Result<Self, Error> {
   |           -                       ^^^^^^^^^^^^^^^^^^^ the trait `WrapperTypeDecode` is not implemented for `contract::Error`, which is required by `Result<ContractRef, contract::Error>: ConstructorReturnType<ContractRef>`
   |  _________|
   | |
14 | |             Ok(Self {})
15 | |         }
   | |_________- required by a bound introduced by this call
   |
   = help: the following other types implement trait `WrapperTypeDecode`:
             Arc<T>
             Box<T>
             Rc<T>
   = note: required for `contract::Error` to implement `ink::parity_scale_codec::Decode`
   = note: required for `Result<ContractRef, contract::Error>` to implement `ConstructorReturnType<ContractRef>`
note: required by a bound in `CreateBuilder::<E, ContractRef, CodeHash, Limits, Endowment, Args, Salt, Unset<ReturnType<()>>>::returns`
  --> $WORKSPACE/crates/env/src/call/create_builder.rs
   |
   |     pub fn returns<R>(
   |            ------- required by a bound in this associated function
...
   |         R: ConstructorReturnType<ContractRef>,
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `CreateBuilder::<E, ContractRef, CodeHash, Limits, Endowment, Args, Salt, Unset<ReturnType<()>>>::returns`

error[E0277]: the trait bound `contract::Error: TypeInfo` is not satisfied
 --> tests/ui/contract/fail/constructor-return-result-non-codec-error.rs:4:16
  |
4 |     pub struct Contract {}
  |                ^^^^^^^^ the trait `TypeInfo` is not implemented for `contract::Error`, which is required by `Result<Result<(), contract::Error>, LangError>: TypeInfo`
  |
  = help: the following other types implement trait `TypeInfo`:
            &T
            &mut T
            ()
            (A, B)
            (A, B, C)
            (A, B, C, D)
            (A, B, C, D, E)
            (A, B, C, D, E, F)
          and $N others
  = note: required for `Result<(), contract::Error>` to implement `TypeInfo`
  = note: 1 redundant requirement hidden
  = note: required for `Result<Result<(), contract::Error>, LangError>` to implement `TypeInfo`
note: required by a bound in `TypeSpec::with_name_str`
 --> $WORKSPACE/crates/metadata/src/specs.rs
  |
  |     pub fn with_name_str<T>(display_name: &'static str) -> Self
  |            ------------- required by a bound in this associated function
  |     where
  |         T: TypeInfo + 'static,
  |            ^^^^^^^^ required by this bound in `TypeSpec::with_name_str`
